llama.cpp

```diff
@@ -7115,15 +7129,22 @@ void llama_sample_grammar(struct llama_context * ctx, llama_token_data_array * c
 
     for (size_t i = 0; i < candidates->size; ++i) {
         const llama_token id    = candidates->data[i].id;
-        const std::string piece = llama_token_to_piece(ctx, id);
+        std::string piece;
+
+        if (pieces != nullptr && pieces[id] != nullptr) {
```

Comment:
The doc comment for this function doesn't clearly state that individual pieces may be nullptr.

Reply:
This is mostly playing defence as std::string cannot handle nullptrs / empty c strings (from my understanding - I'm a complete c++ novice)

The lack of docs allows swapping the behavior from calling `token_to_piece` to assuming a null entry is an empty string, but from my measurements the performance impact was negligible.

If we want to document the current behavior that's fine by me.

Reply:
std::string can represent empty strings, but not null. If it was a C string you would still have to do something with it if it's null to prevent a segfault.

Since this code is apparently not used yet, it's hard to reason about performance or what the correct inputs should be.